File: backend/src/LogistiqueLesLions.Domain/Entities/Contract.cs
using LogistiqueLesLions.Domain.Common;
using LogistiqueLesLions.Domain.Enums;

namespace LogistiqueLesLions.Domain.Entities;

/// <summary>
/// Contrato de compraventa («Contrat de vente») de una negociación.
/// </summary>
/// <remarks>
/// Los datos del vehículo y de las partes se <b>congelan</b> al crear el contrato en
/// lugar de leerse por referencia: un contrato debe reflejar lo acordado en su momento,
/// y si el anuncio cambia de precio o de kilometraje después, el documento firmado no
/// puede cambiar con él.
/// </remarks>
public class Contract : AuditableEntity
{
    /// <summary>Referencia pública única: "YC00125".</summary>
    public string PublicReference { get; set; } = string.Empty;

    public Guid NegotiationId { get; set; }
    public Guid VehicleId { get; set; }
    public Guid SellerId { get; set; }
    public Guid BuyerId { get; set; }

    /// <summary>Quién creó el contrato: la otra parte es quien debe validarlo.</summary>
    public Guid CreatedByUserId { get; set; }

    public ContractStatus Status { get; set; } = ContractStatus.Brouillon;

    // ─── Datos del vehículo, congelados ────────────────────────────────────
    public string VehicleMake { get; set; } = string.Empty;
    public string? VehicleModel { get; set; }
    public string? VehicleVersion { get; set; }
    public int VehicleYear { get; set; }
    public int? VehicleMileage { get; set; }
    public string? VehicleVin { get; set; }
    /// <summary>Matrícula. Suele faltar en el anuncio y se pide al crear el contrato.</summary>
    public string? RegistrationPlate { get; set; }
    /// <summary>Referencia del anuncio de origen.</summary>
    public string VehicleReference { get; set; } = string.Empty;

    // ─── Datos de la operación ─────────────────────────────────────────────
    /// <summary>Precio acordado en FCFA.</summary>
    public decimal AgreedPrice { get; set; }
    public DateTimeOffset SaleDate { get; set; } = DateTimeOffset.UtcNow;

    // ─── Datos legales de las partes ───────────────────────────────────────
    public string SellerLegalName { get; set; } = string.Empty;
    /// <summary>CNI o pasaporte. Se solicita al crear el contrato si no consta.</summary>
    public string? SellerIdDocument { get; set; }
    public string? SellerAddress { get; set; }

    public string BuyerLegalName { get; set; } = string.Empty;
    public string? BuyerIdDocument { get; set; }
    public string? BuyerAddress { get; set; }

    /// <summary>
    /// Código del QR de verificación. Se genera al validar el contrato.
    /// </summary>
    /// <remarks>
    /// Es la credencial que permite comprobar públicamente que el papel que alguien
    /// tiene en la mano corresponde a una venta real: por eso es aleatorio y no
    /// derivado de la referencia, que sí es adivinable.
    /// </remarks>
    public string? VerificationCode { get; set; }

    // ─── Ciclo de vida ─────────────────────────────────────────────────────
    public DateTimeOffset? SentAt { get; set; }
    public DateTimeOffset? ValidatedAt { get; set; }
    public DateTimeOffset? CancelledAt { get; set; }

    /// <summary>Qué pidió corregir quien debía validar.</summary>
    public string? ChangeRequestNotes { get; set; }
    public DateTimeOffset? ChangeRequestedAt { get; set; }

    public Negotiation Negotiation { get; set; } = null!;
    public Vehicle Vehicle { get; set; } = null!;

    /// <summary>Un contrato validado ya no puede modificarse.</summary>
    public bool IsEditable =>
        Status is ContractStatus.Brouillon or ContractStatus.ModificationDemandee;

    /// <summary>Espera la validación de la otra parte.</summary>
    public bool AwaitsValidation => Status == ContractStatus.AValider;

    /// <summary>Quien debe validarlo: siempre la parte que no lo creó.</summary>
    public Guid ValidatorId => CreatedByUserId == SellerId ? BuyerId : SellerId;
}

File: app mobile/lib/features/negotiations/models/contract.dart
// Pestaña «Contrat» de una negociación (`ContractTabDto`).

String contractStatusLabel(String? s) => switch (s) {
      'Brouillon' => 'Brouillon',
      'AValider' => 'À valider',
      'ModificationDemandee' => 'Modification demandée',
      'Valide' => 'Validé',
      'Annule' => 'Annulé',
      _ => '—',
    };

class ContractTab {
  final Contract? contract;
  final ContractPrefill prefill;
  final bool canCreate;
  final bool isSeller;

  const ContractTab({
    required this.prefill,
    required this.canCreate,
    required this.isSeller,
    this.contract,
  });

  factory ContractTab.fromJson(Map<String, dynamic> j) => ContractTab(
        contract: j['contract'] == null
            ? null
            : Contract.fromJson(j['contract'] as Map<String, dynamic>),
        prefill:
            ContractPrefill.fromJson(j['prefill'] as Map<String, dynamic>),
        canCreate: j['canCreate'] as bool? ?? false,
        isSeller: j['isSeller'] as bool? ?? false,
      );
}

class ContractPrefill {
  final String vehicleMake;
  final String? vehicleModel;
  final String? vehicleVersion;
  final int vehicleYear;
  final int? vehicleMileage;
  final String? vehicleVin;
  final String vehicleReference;
  final num suggestedPrice;
  final String sellerLegalName;
  final String buyerLegalName;

  const ContractPrefill({
    required this.vehicleMake,
    required this.vehicleYear,
    required this.vehicleReference,
    required this.suggestedPrice,
    required this.sellerLegalName,
    required this.buyerLegalName,
    this.vehicleModel,
    this.vehicleVersion,
    this.vehicleMileage,
    this.vehicleVin,
  });

  factory ContractPrefill.fromJson(Map<String, dynamic> j) => ContractPrefill(
        vehicleMake: (j['vehicleMake'] ?? '') as String,
        vehicleModel: j['vehicleModel'] as String?,
        vehicleVersion: j['vehicleVersion'] as String?,
        vehicleYear: j['vehicleYear'] as int? ?? 0,
        vehicleMileage: j['vehicleMileage'] as int?,
        vehicleVin: j['vehicleVin'] as String?,
        vehicleReference: (j['vehicleReference'] ?? '') as String,
        suggestedPrice: (j['suggestedPrice'] as num?) ?? 0,
        sellerLegalName: (j['sellerLegalName'] ?? '') as String,
        buyerLegalName: (j['buyerLegalName'] ?? '') as String,
      );
}

class Contract {
  final String id;
  final String publicReference;
  final String status;
  final String negotiationId;

  final String vehicleMake;
  final String? vehicleModel;
  final String? vehicleVersion;
  final int vehicleYear;
  final int? vehicleMileage;
  final String? vehicleVin;
  final String? registrationPlate;
  final String vehicleReference;

  final num agreedPrice;
  final DateTime saleDate;

  final String sellerLegalName;
  final String? sellerIdDocument;
  final String? sellerAddress;
  final String buyerLegalName;
  final String? buyerIdDocument;
  final String? buyerAddress;

  final bool createdByMe;
  final bool canEdit;
  final bool canSend;
  final bool canValidate;
  final bool canRequestChanges;
  final bool canCancel;
  final bool canDownloadPdf;

  final String? verificationCode;
  final String? changeRequestNotes;
  final DateTime createdAt;
  final DateTime? sentAt;
  final DateTime? validatedAt;

  const Contract({
    required this.id,
    required this.publicReference,
    required this.status,
    required this.negotiationId,
    required this.vehicleMake,
    required this.vehicleYear,
    required this.vehicleReference,
    required this.agreedPrice,
    required this.saleDate,
    required this.sellerLegalName,
    required this.buyerLegalName,
    required this.createdByMe,
    required this.canEdit,
    required this.canSend,
    required this.canValidate,
    required this.canRequestChanges,
    required this.canCancel,
    required this.canDownloadPdf,
    required this.createdAt,
    this.vehicleModel,
    this.vehicleVersion,
    this.vehicleMileage,
    this.vehicleVin,
    this.registrationPlate,
    this.sellerIdDocument,
    this.sellerAddress,
    this.buyerIdDocument,
    this.buyerAddress,
    this.verificationCode,
    this.changeRequestNotes,
    this.sentAt,
    this.validatedAt,
  });

  factory Contract.fromJson(Map<String, dynamic> j) => Contract(
        id: j['id'] as String,
        publicReference: (j['publicReference'] ?? '') as String,
        status: (j['status'] ?? 'Brouillon') as String,
        negotiationId: (j['negotiationId'] ?? '') as String,
        vehicleMake: (j['vehicleMake'] ?? '') as String,
        vehicleModel: j['vehicleModel'] as String?,
        vehicleVersion: j['vehicleVersion'] as String?,
        vehicleYear: j['vehicleYear'] as int? ?? 0,
        vehicleMileage: j['vehicleMileage'] as int?,
        vehicleVin: j['vehicleVin'] as String?,
        registrationPlate: j['registrationPlate'] as String?,
        vehicleReference: (j['vehicleReference'] ?? '') as String,
        agreedPrice: (j['agreedPrice'] as num?) ?? 0,
        saleDate:
            DateTime.tryParse((j['saleDate'] ?? '') as String) ?? DateTime.now(),
        sellerLegalName: (j['sellerLegalName'] ?? '') as String,
        sellerIdDocument: j['sellerIdDocument'] as String?,
        sellerAddress: j['sellerAddress'] as String?,
        buyerLegalName: (j['buyerLegalName'] ?? '') as String,
        buyerIdDocument: j['buyerIdDocument'] as String?,
        buyerAddress: j['buyerAddress'] as String?,
        createdByMe: j['createdByMe'] as bool? ?? false,
        canEdit: j['canEdit'] as bool? ?? false,
        canSend: j['canSend'] as bool? ?? false,
        canValidate: j['canValidate'] as bool? ?? false,
        canRequestChanges: j['canRequestChanges'] as bool? ?? false,
        canCancel: j['canCancel'] as bool? ?? false,
        canDownloadPdf: j['canDownloadPdf'] as bool? ?? false,
        verificationCode: j['verificationCode'] as String?,
        changeRequestNotes: j['changeRequestNotes'] as String?,
        createdAt:
            DateTime.tryParse((j['createdAt'] ?? '') as String) ?? DateTime.now(),
        sentAt: DateTime.tryParse((j['sentAt'] ?? '') as String),
        validatedAt: DateTime.tryParse((j['validatedAt'] ?? '') as String),
      );
}
